### PR TITLE
Add filtering API to LazyDataView

### PR DESCRIPTION
**Pull request by mshabarov**
Part of #8646 

New interface `FilterableLazyDataView` has been added to provide an API for setting the filter to lazy data load Components. Main target for this - ComboBox.

### CODE REVIEW

**Comment by mshabarov** [comment]
Converted to Draft because this API needs a design review.

File: flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractFilterableLazyDataView.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    /**
+     * Creates a new instance, verifies the passed data provider is compatible
+     * with this data view implementation, and sets a callbacks for handling the
+     * items filtering.
+     *
+     * @param dataCommunicator
+     *            the data communicator of the component
+     * @param component
+     *            the component
+     * @param filterConsumer
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractFilterableLazyDataView.java:53]: done

File: flow-data/src/main/java/com/vaadin/flow/data/provider/FilterCombiner.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.vaadin.flow.data.provider;
+
+import com.vaadin.flow.function.SerializableBiFunction;
+
+/**
+ * A callback interface that is used to combine a new configured filter (given
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/FilterCombiner.java]: done

File: flow-data/src/main/java/com/vaadin/flow/data/provider/FilterableLazyDataView.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    default void removeFilter() {
+        setFilter(null);
+    }
+
+    /**
+     * Sets the filter combiner function which will be used to merge the
+     * previous configured filter and a new filter set with
+     * {@link #setFilter(Object)}.
+     * 
+     * @param filterCombiner
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/FilterableLazyDataView.java:57]: done

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasFilterableLazyDataView.java [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+package com.vaadin.flow.data.provider;
+
+import java.io.Serializable;
+import java.util.Collection;
+
+import com.vaadin.flow.function.SerializableFunction;
+
+/**
+ * Interface that defines methods for fetching items lazily from a backend
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/HasFilterableLazyDataView.java]: done

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasFilterableLazyDataView.java [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+     * <p>
+     * {@code component.setItemsWithFilter(query -> orderService
+     *      .getOrders(query.getOffset(), query.getLimit(), query.getFilter());}
+     * <p>
+     * The returned data view object can be used for further configuration, or
+     * later on fetched with {@link #getFilterableLazyDataView()}. For using
+     * in-memory data, like {@link java.util.Collection}, use
+     * {@link HasListDataView#setItems(Collection)} instead.
+     *
+     * @param fetchCallback
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/HasFilterableLazyDataView.java:60]: done

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasFilterableLazyDataView.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+     *                    query -> orderService.getOrders(query.getOffset, 
+     *                                     query.getLimit(), query.getFilter()),
+     *                    query -> orderService.getSize(query.getFilter()));}
+     * <p>
+     * The returned data view object can be used for further configuration, or
+     * later on fetched with {@link #getFilterableLazyDataView()}. For using
+     * in-memory data, like {@link java.util.Collection}, use
+     * {@link HasListDataView#setItems(Collection)} instead.
+     *
+     * @param fetchCallback
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/HasFilterableLazyDataView.java:97]: done

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasFilterableLazyDataView.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+     * {@link #setItemsWithFilter(CallbackDataProvider.FetchCallback)} or
+     * {@link #setItemsWithFilter(CallbackDataProvider.FetchCallback, SerializableFunction)}
+     * instead.
+     * <p>
+     * The returned data view object can be used for further configuration, or
+     * later on fetched with {@link #getFilterableLazyDataView()}. For using
+     * in-memory data, like {@link java.util.Collection}, use
+     * {@link HasListDataView#setItems(Collection)} instead.
+     *
+     * @param dataProvider
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/HasFilterableLazyDataView.java:183]: done

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasFilterableLazyDataView.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+     *                    query -> orderService.getOrders(query.getOffset,
+     *                                     query.getLimit(), query.getFilter()),
+     *                    query -> orderService.getSize(query.getFilter()));}
+     * <p>
+     * The returned data view object can be used for further configuration, or
+     * later on fetched with {@link #getFilterableLazyDataView()}. For using
+     * in-memory data, like {@link java.util.Collection}, use
+     * {@link HasListDataView#setItems(Collection)} instead.
+     *
+     * @param fetchCallback
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/HasFilterableLazyDataView.java:214]: done

File: flow-data/src/test/java/com/vaadin/flow/data/provider/AbstractFilterableLazyDataViewTest.java [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+            if (!Strings.isNullOrEmpty(lastName)) {
+                result &= lastName.equalsIgnoreCase(person.getLastName());
+            }
+            if (age > 0) {
+                result &= age == person.getAge();
+            }
+            return result;
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-data/src/test/java/com/vaadin/flow/data/provider/AbstractFilterableLazyDataViewTest.java]: done

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasFilterableLazyDataView.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (denis-anisimov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/HasFilterableLazyDataView.java:60]: I don't get from javadocs the relation between `setItemsWithConvertedFilter` and `setItemsWithFilter`.
Are they mutually exclusive ?
What will be the effect if I call:
```
setItemsWithFilter
setItemsWithConvertedFilter
```
?
Will the second call completely cancel the first one ?
Or I they work independent ?
This is extremely confusing thing when several methods mutate the same state.
If they mutate the same state then they should have the same name : so it should be overloads.
Even overloads are confusing. But at least they are less confusing.
At least javadocs should mention that they mutates the same state and any subsequent call of other method 
completely cancels the effect of the previous one.
And all methods should have crossreferences to each other via `@see` (to all overloads/any mutator the same state).

**Comment by pleku** [comment]
I think this can be now closed @mshabarov ?

**Comment by mshabarov** [comment]
@pleku Yes, let's close this one as we conclude the API design and create a new fresh PR later-on.

**Pull request closed without merging**